pkg/security/security_profile/activity_tree/activity_tree.go

```diff
@@ -0,0 +1,381 @@
+}
+
+// isEventValid evaluates if the provided event is valid
+func (at *ActivityTree) isEventValid(event *model.Event, shadowInsertion bool) (bool, error) {
+	// check event type
+	if !at.validator.IsEventTypeValid(event.GetEventType()) {
+		if !shadowInsertion {
+			at.Stats.droppedCount[event.GetEventType()][eventTypeReason].Inc()
+		}
+		return false, fmt.Errorf("event type not valid: %s", event.GetEventType())
+	}
+
+	// drop events with an error
+	if event.Error != nil {
+		if !shadowInsertion {
+			at.Stats.droppedCount[event.GetEventType()][eventTypeReason].Inc()
+		}
+		return false, fmt.Errorf("event error: %w", event.Error)
+	}
+
+	// event specific filtering
+	switch event.GetEventType() {
+	case model.BindEventType:
+		// ignore non IPv4 / IPv6 bind events for now
+		if event.Bind.AddrFamily != unix.AF_INET && event.Bind.AddrFamily != unix.AF_INET6 {
+			if !shadowInsertion {
+				at.Stats.droppedCount[model.BindEventType][bindFamilyReason].Inc()
+			}
+			return false, fmt.Errorf("invalid bind family")
+		}
+	}
+	return true, nil
+}
+
+// Insert inserts the event in the activity tree
+func (at *ActivityTree) Insert(event *model.Event, generationType NodeGenerationType) (bool, error) {
+	return at.insert(event, false, generationType)
+}
+
+// Contains looks up the event in the activity tree
+func (at *ActivityTree) Contains(event *model.Event, generationType NodeGenerationType) (bool, error) {
+	newEntry, err := at.insert(event, true, generationType)
+	return !newEntry, err
+}
+
+// insert inserts the event in the activity tree, returns true if the event generated a new entry in the tree
+func (at *ActivityTree) insert(event *model.Event, shadowInsertion bool, generationType NodeGenerationType) (newEntry bool, outputErr error) {
+	// sanity check
+	if generationType > SecurityProfile {
+		return false, fmt.Errorf("invalid generation type: %v", generationType)
+	}
+
+	// metrics
+	defer func() {
+		if newEntry && !shadowInsertion {
+			// this doesn't count the exec events which are counted separately
+			at.Stats.addedCount[event.GetEventType()][generationType].Inc()
+		}
+	}()
+
+	// check if this event type is traced
+	if valid, err := at.isEventValid(event, shadowInsertion); !valid || err != nil {
+		return false, fmt.Errorf("invalid event: %s", err)
+	}
+
+	// find the node where the event should be inserted
+	entry, isResolved := event.FieldHandlers.ResolveProcessCacheEntry(event)
+	if !isResolved {
+		return false, fmt.Errorf("process cache entry not found")
+	}
+	if !entry.HasCompleteLineage() && !shadowInsertion { // check that the process context lineage is complete, otherwise drop it
+		at.Stats.droppedCount[event.GetEventType()][brokenLineageReason].Inc()
+		return false, fmt.Errorf("incomplete lineage")
+	}
+
+	node, newProcessNode, err := at.CreateProcessNode(entry, generationType, shadowInsertion)
```

Comment:
IMO we should only CreateProcessNode if !shadowInsertion

Reply:
Actually this call is needed because we use it to go up the process tree and check the process lineage ... but I'll think again about not calling `New*Node` ... as you suggested later we might not need it.